File: integration-test/src/test/js/steps-token-types.js
const { strictEqual } = require('assert');
const { After, Given, Then, When } = require('@cucumber/cucumber');
const {
    clientRegistersTokenType,
    thereIsAClient,
    tokenTypesContextMixin,
    walletContextMixin,
} = require('./cnft-common-steps');
const { Ascii, log } = require('./cnft-util');

const { FgGreen: FgBase, FgMagenta: FgVal, Reset } = Ascii;

const context = {};
Object.assign(context, walletContextMixin, tokenTypesContextMixin);

Given(
    '[Token Types] There is a client {string}',
    async function (clientName) {
        await thereIsAClient(context, clientName);
    }
);

When(
    '[Token Types] Client {string} registers token type {string}',
    async function (clientName, tokenTypeName) {
        await clientRegistersTokenType(context, clientName, tokenTypeName);
    }
);

Then(
    '[Token Types] Client {string} sees {int} token types in his list',
    async function (clientName, expectedTokenTypesCount) {
        log(`${FgBase}Client ${FgVal}'${clientName}'${FgBase} sees ${FgVal}${expectedTokenTypesCount}` +
            `${FgBase} token types in his list${Reset}`);

        const wallet = context.getWalletByClientName(clientName);
        const actualTokenTypes = await wallet.listOwnedTokenTypes;

        strictEqual(actualTokenTypes.length, expectedTokenTypesCount);

        log(`${FgBase}  Done: correct!${Reset}\n`);
    }
);

After(function () {
/*    return new Promise((resolve, _reject) => {
        log(`Stopping wallets ...`);
        for (const index in context.wallets) {
            log(`Stopping wallet ${index} ...`);
        }
        log(`Wallets stopped.`);
        return resolve();
    });*/
});

File: integration-test/src/test/scala/ru/sberbank/blockchain/cnft/wallet/test/StepsTokenTypes.scala
package ru.sberbank.blockchain.cnft.wallet.test

import io.cucumber.scala.{EN, ScalaDsl}
import org.slf4j.LoggerFactory
import ru.sberbank.blockchain.cnft.wallet.dsl._
import ru.sberbank.blockchain.cnft.wallet.test.CNFTCommonSteps.{TokenTypesContext, WalletsContext, clientRegistersTokenType, thereIsAClient}

class StepsTokenTypes extends ScalaDsl with EN {

    private val log = LoggerFactory.getLogger(classOf[StepsTokenTypes])

    //

//    After { _: Scenario =>
//        log.info("Stopping wallets ...")
//        context.wallets.values.foreach(_.stopListenBlocks())
//        log.info("All wallets stopped.")
//    }

    Given("""[Token Types] There is a client {string}""") { clientName: String =>
        log.info(s"Creating a new wallet for client '$clientName'...")
        thereIsAClient(context, clientName)
        log.info(s"Created a new wallet for client '$clientName'")
    }

    When("""[Token Types] Client {string} registers token type {string}""") {
        (clientName: String, tokenTypeName: String) =>
            log.info(s"Client '$clientName' registering token type '$tokenTypeName'...")
            clientRegistersTokenType(
                context,
                clientName,
                tokenTypeName
            )
            context.withTokenType(tokenTypeName) { tokenTypeId =>
                log.info(s"Client '$clientName' registered token type '$tokenTypeName' ($tokenTypeId)")
            }
    }

    Then("""[Token Types] Client {string} sees {int} token types in his list""") {
        (clientName: String, expectedTokenTypesCount: Int) =>
            log.info(s"Checking client '$clientName' has $expectedTokenTypesCount token types in his list...")
            context.withWallet(clientName) { wallet =>
                val actualTokenTypes =
                    wallet
                        .listOwnedTokenTypes
                        .orFail("Failed to check token types")
                assert(actualTokenTypes.length == expectedTokenTypesCount)
            }
    }

    //

    private val _context = new ThreadLocal[ScenarioContext]() {
        override def initialValue(): ScenarioContext = new ScenarioContext
    }

    private def context: ScenarioContext = _context.get()

    class ScenarioContext extends WalletsContext with TokenTypesContext

}
